dcc-business-rules-utils/src/validate.ts

```diff
@@ -0,0 +1,110 @@
+import { ErrorObject } from "ajv"
+import { specificationVersion } from "certlogic-js"
+import { dateFromString } from "certlogic-js/dist/internals"
+import { dataAccesses, validateFormat, ValidationError } from "certlogic-js/dist/validation"
+import { gt } from "semver"
+
+import { createSchemaValidator } from "./schema-validator"
+import { CertificateType, Rule } from "./typings"
+
+
+const ruleSchemaValidator = createSchemaValidator(require("./validation-rule.schema.json"))
+
+
+const areEqual = (leftSet: string[], rightSet: string[]): boolean =>
+    leftSet.length === rightSet.length && leftSet.every((item) => rightSet.indexOf(item) > -1)
+
+export type AffectedFieldsValidationResult = null | { actual: string[], computed: string[] }
+
+const validateAffectedFields = (rule: any): AffectedFieldsValidationResult => {
+    const payloadPrefix = "payload."
+    const actual = rule.AffectedFields
+    const computed = dataAccesses(rule.Logic)
+        .filter((fieldName) => fieldName.startsWith(payloadPrefix))
+        .map((fieldName) => fieldName.substring(payloadPrefix.length))
+    return areEqual(actual, computed)
+        ? null
+        : { actual, computed }
+}
+
+
+const validateMetaData = (rule: Rule): string[] => {
+    const errors: string[] = []
+
+    const joinWrapped = (strings: string[]) => strings.map((str) => `"${str}"`).join(", ")
+
+    const ruleTypes = [ "Acceptance", "Invalidation" ]
+    if (ruleTypes.indexOf(rule.Type) === -1) {
+        errors.push(`Type "${rule.Type}" must be one of [ ${joinWrapped(ruleTypes)} ]`)
+    }
+    if (rule.Engine !== "CERTLOGIC") {
+        errors.push(`Engine "${rule.Engine}" must be "CERTLOGIC"`)
+    }
+    if (gt(rule.EngineVersion, specificationVersion)) {
+        errors.push(`EngineVersion ${rule.EngineVersion} is newer than the currently supported version ${specificationVersion}`)
+    }
+    if (gt("0.7.5", rule.EngineVersion)) {
+        errors.push(`EngineVersion ${rule.EngineVersion} must be 0.7.5 or newer`)
+    }
+    if (!rule.ValidFrom) {
+        errors.push(`ValidFrom must be defined`)
+    } else {
+        const validFrom = dateFromString(rule.ValidFrom)
+        if (rule.ValidTo) {
+            const validTo = dateFromString(rule.ValidTo)
+            if (validFrom > validTo) {
+                errors.push(`ValidFrom ${rule.ValidFrom} must be before ValidTo ${rule.ValidTo}`)
+            }
+            let validFromPlus72hours = validFrom
+            validFrom.setUTCHours(validFromPlus72hours.getUTCHours() + 72)
```

Comment:
This logic seems a bit flawed. You create a new variable, but add hours to the old one. (This will will work in Dart since you created a reference rather than a copy, but I'm not sure in ts/js 🤔)

Reply:
Good find! Fixing that.

Reply:
Committed + pushed a fix.